users/postgres/users.go

```diff
@@ -141,6 +141,95 @@ func (ur userRepository) RetrieveByID(ctx context.Context, id string) (users.Use
 	return toUser(dbu)
 }
 
+func (ur userRepository) Users(ctx context.Context, offset, limit uint64, email string, um users.Metadata) (users.UserPage, error) {
+	emq, mb, err := createUsersListQuery(email, um)
+	if err != nil {
+		return users.UserPage{}, errors.Wrap(errRetrieveDB, err)
+	}
+
+	// Create LIKE operator to search Users with email containing a given string
+	if email != "" {
+		email = fmt.Sprintf(`%%%s%%`, email)
+	}
+
+	q := fmt.Sprintf(`SELECT id, email, metadata FROM users %s ORDER BY email LIMIT :limit OFFSET :offset;`, emq)
+
+	params := map[string]interface{}{
+		"limit":    limit,
+		"offset":   offset,
+		"email":    email,
+		"metadata": mb,
+	}
+
+	rows, err := ur.db.NamedQueryContext(ctx, q, params)
+	if err != nil {
+		return users.UserPage{}, errors.Wrap(errSelectDb, err)
+	}
+	defer rows.Close()
+
+	var items []users.User
+	for rows.Next() {
+		dbusr := dbUser{}
+		if err := rows.StructScan(&dbusr); err != nil {
+			return users.UserPage{}, errors.Wrap(errSelectDb, err)
+		}
+
+		user, err := toUser(dbusr)
+		if err != nil {
+			return users.UserPage{}, err
+		}
+
+		items = append(items, user)
+	}
+
+	cq := fmt.Sprintf(`SELECT COUNT(*) FROM users %s;`, emq)
+
+	total, err := total(ctx, ur.db, cq, params)
+	if err != nil {
+		return users.UserPage{}, errors.Wrap(errSelectDb, err)
+	}
+
+	page := users.UserPage{
+		Users: items,
+		PageMetadata: users.PageMetadata{
+			Total:  total,
+			Offset: offset,
+			Limit:  limit,
+		},
+	}
+
+	return page, nil
+}
+
+func createUsersListQuery(email string, m users.Metadata) (string, []byte, error) {
+	q := ""
+	qMeta := []byte("{}")
+
+	if email == "" && len(m) == 0 {
+		return q, qMeta, nil
+	}
+
+	if email != "" && len(m) == 0 {
+		return q, qMeta, nil
+	}
+
+	if len(m) > 0 {
+		b, err := json.Marshal(m)
+		if err != nil {
+			return "", []byte{}, err
+		}
+		qMeta = b
+	}
```

Comment:
Can you generalize this, [getUsersMetadataQuery](https://github.com/mainflux/mainflux/blob/420b598ac711e56e6d4d86987f8dd0e1caeb9532/users/postgres/users.go#L293), and [getGroupsMetadataQuery](https://github.com/mainflux/mainflux/blob/420b598ac711e56e6d4d86987f8dd0e1caeb9532/users/postgres/groups.go#L427) to a single parameterized method?